src/main/src/main/java/org/geoserver/security/auth/EnvPwdResolverUserGroupServiceWrapper.java

```diff
@@ -0,0 +1,95 @@
+/* (c) 2022 Open Source Geospatial Foundation - all rights reserved
+ * This code is licensed under the GPL 2.0 license, available at the root
+ * application directory.
+ */
+package org.geoserver.security.auth;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.apache.commons.lang.StringUtils;
+import org.geoserver.platform.GeoServerEnvironment;
+import org.geoserver.platform.GeoServerExtensions;
+import org.geoserver.security.GeoServerSecurityManager;
+import org.geoserver.security.GeoServerUserGroupService;
+import org.geoserver.security.impl.GeoServerUser;
+import org.geoserver.security.password.GeoServerPasswordEncoder;
+import org.geotools.util.logging.Logging;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+
+/**
+ * A UserGroupService wrapper able to resolve password that have been parametrized. The placeholder
+ * can be encoded in plain txt or any reversible encoding.
+ */
+class EnvPwdResolverUserGroupServiceWrapper implements UserDetailsService {
+
+    private GeoServerUserGroupService delegate;
+    private final GeoServerEnvironment environment;
+
+    private static Logger LOGGER = Logging.getLogger(EnvPwdResolverUserGroupServiceWrapper.class);
+
+    EnvPwdResolverUserGroupServiceWrapper(GeoServerUserGroupService userDetailsService) {
+        this.delegate = userDetailsService;
+        environment = GeoServerExtensions.bean(GeoServerEnvironment.class);
+    }
+
+    @Override
+    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {
+        UserDetails userDetails = delegate.loadUserByUsername(username);
+        if (environment != null && GeoServerEnvironment.allowEnvParametrization()) {
```

Comment:
Since this can be activated only at startup time, I'd suggest to move the check into ``UsernamePasswordAuthenticationProvider`` and just skip the wrapping if env parametrization is not enabled or not available.

Reply:
fixed